RIPS/rip-9997.md

```diff
@@ -0,0 +1,204 @@
+---
+rip: 0
+title: An RIP-7560 transactions bundle subtype
+description: An RIP-7560 transaction subtype that allows specifying a set of AA transactions that can be atomically validated together
+author: Vitalik Buterin (@vbuterin), Yoav Weiss (@yoavw), Alex Forshtat (@forshtat), Dror Tirosh (@drortirosh), Shahaf Nacson (@shahafn)
+discussions-to: https://ethereum-magicians.org/t/rip-7560-native-account-abstraction/16664
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-09-01
+requires: 7560
+---
+
+
+## Abstract
+
+This proposal provides block builders with a mechanism to isolate validation from execution for
+a set of RIP-7560 transactions.
+By doing so we simplify the task of filling a block gas space with RIP-7560 transactions and prevent potential
+denial of service attacks against block builders.
+
+On "single sequencer" Layer 2 chains that do not have a "transaction mempool" in a traditional sense,
+this proposal provides no benefit compared to the original RIP-7560.
+It is required for DoS mitigation on every chain that does rely on a public mempool for block building, however.
+
+## Motivation
+
+The `AA_TX_TYPE` transaction type that is defined in RIP-7560 completely separates transaction validity
+on a protocol level from ECDSA signatures or any other protocol-defined rules, instead allowing account's EVM code
+to determine a validity of a transaction.
+
+This validation EVM code, however, may be affected by any observable state change within a block.
+A block builder trying to include a set of `AA_TX_TYPE` transactions may face a challenge of finding a combination
+that fits together without invalidating each other.
+
+The block builder has to execute the entire body of one transaction before picking the next one.
+Additionally, it is not feasible to apply any limitations on the transaction execution, while it is expected that
+the validation code will be constrained by the block builder with a ruleset like ERC-7562.
+
+This document proposes a mechanism for block builder to specify a set of `AA_TX_TYPE` transactions that are guaranteed
+to fit together due to a fact that all of their validation frames are executed consecutively first,
+before their respective execution frames.
+
+## Specification
+
+### Bundle header RIP-7560 transaction subtype
+
+```
+BUNDLE_HEADER_SUBTYPE = x
+```
+
+Instead, we propose to introduce an explicit "counter" transaction subtype.
+
+The payload should be interpreted as:
+
+```
+0x04 || 0x01 || rlp([chainId, transactionCount])
+```
+
+Header transactions have a unique hash calculated as follows:
+
+```
+keccak256(AA_TX_TYPE || 0x01 || rlp(chainId, transactionCount, blockNumber, txIndex))
```

Comment:
why do you pack txIndex, blockNumber into rlp ? this requires the node to decode the RLP-encoded TX, and re-create an elp block, just to hash.
Instead, add it as appended fixed-size parameters, after `|| 0x01 ||`

Reply:
@drortirosh I am not sure I understand your description of the problem. Please elaborate why appending `blockNumber||txIndex` is superior.

Reply:
@drortirosh ping on this one. Please help.